### ISSUE TITLE
Migrate neural net agents across AlphaZero.jl instances?

### ISSUE THREAD

**Issue by smart-fr** [jonathan-laurent/AlphaZero.jl#34]
I would like to train an agent on a cloud-based multi-GPU platform and then migrate it to a much cheaper platform used only for inference. The first migration step would be towards a secondary instance of AlphaZero.jl, assuming that the resources needed for testing the quality of the agent may be greatly reduced compared with the resources needed to train it.

I have two related questions:
1. Since AlphaZero.jl was designed with accessibility in mind, so that it performs very well on a desktop computer, is there a real benefit from running it on a multi-GPU platform and if yes, is there some parameter tweaking to do in order to leverage the extended hardware?
2. Considering an agent has been trained on a given instance of AlphaZero.jl, is there a "simple" way to migrate it into another instance of AlphaZero.jl?

**Comment by jonathan-laurent** [comment]
> Since AlphaZero.jl was designed with accessibility in mind, so that it performs very well on a desktop computer, is there a real benefit from running it on a multi-GPU platform and if yes, is there some parameter tweaking to do in order to leverage the extended hardware?

Right now, only the self-play phase is parallelized and can take advantage of multiple machines / GPUs. This is usually the bottleneck though so you can probably take advantage of up to 10 machines (see the JuliaCon [sponsor talk](https://www.youtube.com/watch?v=JVUJ5Oohuhs) that did this) if you use some choice of hyperparameters where you run few iterations that simulate a lot of games each.

> Considering an agent has been trained on a given instance of AlphaZero.jl, is there a "simple" way to migrate it into another instance of AlphaZero.jl?

Yes, just copy the `session` folder.

**Comment by smart-fr** [comment]
Very clear, thank you!

**Comment by smart-fr** [comment]
> Right now, only the self-play phase is parallelized and can take advantage of multiple machines / GPUs.

This means that even to benefit from several CPUs outside of the self-play phase, one would need to refactor some AlphaZero.jl code to parallelize computations? This parallelization doesn't currently occur "just" because we are using Julia / AlphaZero.jl?
So, there's no need to stack vCPUs in the platform in order to speed up AlphaZero.jl processing?

**Comment by jonathan-laurent** [comment]
On a single machine, most of the runtime happens during self-play (~90% for the default hyperparameters) so there is still a good opportunity to parallelize. 

Ultimately, I agree that some refactoring would be needed to leverage more parallelism, but I do not want to do this at the expense of simplicity.

**Comment by smart-fr** [comment]
Indeed.
This [JuliaCon sponsor talk](https://www.youtube.com/watch?v=JVUJ5Oohuhs) is brilliant!
